File: tools/flightfinder/flightfinder/consensus.py
"""Modo consenso: cruza as fontes, reconcilia preços (em BRL) e dá confiança.

Para cada fonte pegamos a tarifa mais barata comparável (já em BRL), calculamos
a faixa (mín/mediana/máx) e a dispersão. Sinalizamos divergência quando a
dispersão passa de um limiar (padrão 15%).
"""
from __future__ import annotations

from dataclasses import dataclass, field
from statistics import median

from .model import Offer, SourceResult

DIVERGENCE_THRESHOLD = 0.15


@dataclass
class Consensus:
    sources_used: list[str] = field(default_factory=list)
    sources_failed: list[str] = field(default_factory=list)
    cheapest_offer: Offer | None = None
    min_brl: float | None = None
    median_brl: float | None = None
    max_brl: float | None = None
    spread_pct: float | None = None
    confidence: str = "indisponível"
    price_signal: str | None = None
    divergent: bool = False
    notes: list[str] = field(default_factory=list)


def _signal_pt(sig: str | None) -> str | None:
    return {"low": "barata", "typical": "típica", "high": "cara"}.get(sig or "")


def build(results: list[SourceResult], threshold: float = DIVERGENCE_THRESHOLD) -> Consensus:
    c = Consensus()
    per_source: list[float] = []
    all_priced: list[Offer] = []

    for r in results:
        if r.error:
            c.sources_failed.append(f"{r.name} ({r.error})")
            continue
        if r.price_signal and not c.price_signal:
            c.price_signal = r.price_signal
        cheap = r.cheapest_brl()
        if cheap is None:
            # fonte respondeu mas nada pôde ser convertido para BRL
            c.sources_failed.append(f"{r.name} (sem preço convertível)")
            continue
        c.sources_used.append(r.name)
        per_source.append(cheap.price_brl)
        all_priced.extend(r.priced())

    # melhor opção: a mais barata que tenha companhia identificada (mais útil);
    # se nenhuma tiver, cai para a mais barata no geral.
    named = [o for o in all_priced if (o.airline or "").strip()]
    pool = named or all_priced
    if pool:
        c.cheapest_offer = min(pool, key=lambda o: o.price_brl)

    n = len(per_source)
    if n:
        c.min_brl, c.max_brl = min(per_source), max(per_source)
        c.median_brl = median(per_source)
        if c.min_brl > 0:
            c.spread_pct = (c.max_brl - c.min_brl) / c.min_brl
    c.divergent = c.spread_pct is not None and c.spread_pct > threshold

    if n >= 2 and not c.divergent:
        c.confidence = "alta"
    elif n >= 2 and c.spread_pct is not None and c.spread_pct <= 0.30:
        c.confidence = "média"
    elif n >= 2:
        c.confidence = "baixa"
    elif n == 1:
        c.confidence = "baixa"
    else:
        c.confidence = "indisponível"

    sig = _signal_pt(c.price_signal)
    if sig:
        c.notes.append(f"Para esta rota e data, a tarifa está {sig} (sinal fast-flights).")
    if c.divergent:
        c.notes.append(
            f"Fontes divergem {c.spread_pct * 100:.0f}% (acima de {threshold * 100:.0f}%): "
            "pode haver combos de autoconexão (bilhetes separados); confira as condições."
        )
    return c

File: lib/consensus.ts
// Modo consenso HONESTO.
// Regra inviolável: dado fabricado (demo) nunca entra em ranking nem dá confiança.
// Regra de independência: fast-flights e google-flights são o MESMO motor (Google)
// e contam UMA vez. Confiança "alta" exige >=2 motores independentes REAIS (live).
import type { Consensus, DataQuality, Offer } from "@/lib/types";

export interface SourceResult {
  name: string;
  engine: string;
  dataQuality: DataQuality;
  offers: Offer[];
  priceSignal?: "low" | "typical" | "high" | null;
  error?: string | null;
}

const DIVERGENCE_THRESHOLD = 0.15;
const isReal = (q: DataQuality) => q === "live" || q === "cache";

function median(xs: number[]): number {
  const s = [...xs].sort((a, b) => a - b);
  const m = Math.floor(s.length / 2);
  return s.length % 2 ? s[m] : (s[m - 1] + s[m]) / 2;
}
function signalPt(sig?: string | null): string | null {
  return sig === "low" ? "barata" : sig === "typical" ? "típica" : sig === "high" ? "cara" : null;
}

export function buildConsensus(results: SourceResult[], threshold = DIVERGENCE_THRESHOLD): Consensus {
  const c: Consensus = {
    sourcesUsed: [], independentEngines: [], sourcesFailed: [],
    cheapest: null, minBrl: null, medianBrl: null, maxBrl: null, spreadPct: null,
    confidence: "indisponível", priceSignal: null, divergent: false, dataQuality: "unavailable", notes: [],
  };

  for (const r of results) {
    if (r.error || !r.offers.length) {
      if (r.error) c.sourcesFailed.push(`${r.name} (${r.error})`);
      continue;
    }
  }

  // ofertas reais (live ou cache), com preço em BRL. Demo NUNCA entra.
  const real = results.flatMap((r) => r.offers).filter((o) => isReal(o.dataQuality) && o.priceBrl != null);
  const live = real.filter((o) => o.dataQuality === "live");

  // sinal de preço só de fonte real que de fato computou (fast-flights = google).
  const signalSrc = results.find((r) => isReal(r.dataQuality) && r.priceSignal && r.offers.length);
  c.priceSignal = signalSrc?.priceSignal ?? null;

  // dedupe por MOTOR: menor preço por motor (google une fast-flights + google-flights).
  const byEngine = new Map<string, Offer>();
  for (const o of real) {
    const cur = byEngine.get(o.engine);
    if (!cur || (o.priceBrl as number) < (cur.priceBrl as number)) byEngine.set(o.engine, o);
  }
  const perEngine = [...byEngine.values()];
  const liveEngines = [...new Set(live.map((o) => o.engine))];

  c.independentEngines = [...byEngine.keys()];
  c.sourcesUsed = [...new Set(real.map((o) => o.source))];

  if (perEngine.length) {
    const prices = perEngine.map((o) => o.priceBrl as number);
    c.minBrl = Math.min(...prices);
    c.maxBrl = Math.max(...prices);
    c.medianBrl = median(prices);
    if (c.minBrl > 0) c.spreadPct = (c.maxBrl - c.minBrl) / c.minBrl;
    c.cheapest = (live.length ? live : real).reduce((a, b) => ((a.priceBrl as number) <= (b.priceBrl as number) ? a : b));
  }
  c.divergent = c.spreadPct != null && c.spreadPct > threshold;
  const anyDemo = results.some((r) => r.offers.some((o) => o.dataQuality === "demo"));
  c.dataQuality = live.length ? "live" : real.length ? "cache" : anyDemo ? "demo" : "unavailable";

  if (!real.length) {
    c.confidence = "indisponível";
  } else if (liveEngines.length >= 2 && !c.divergent) {
    c.confidence = "alta";
  } else if (liveEngines.length >= 2) {
    c.confidence = "média";
  } else {
    c.confidence = "baixa"; // 1 motor live, ou só cache
  }

  // notas honestas
  const sig = signalPt(c.priceSignal);
  if (sig) c.notes.push(`Para esta rota e data, a tarifa está ${sig} (sinal do motor Google via fast-flights).`);
  const engines = new Set(real.map((o) => o.engine));
  if (engines.has("google")) c.notes.push("fast-flights e google-flights usam o mesmo motor (Google): contam como uma fonte.");
  if (liveEngines.length < 2)
    c.notes.push("Confiança limitada: menos de 2 fontes independentes reais. Configure mais fontes (worker, Duffel, seats.aero) para validar.");
  if (!live.length && real.length) c.notes.push("Preço apenas de cache (Travelpayouts), sem validação live: trate como referência, não como preço final.");
  if (c.divergent && c.spreadPct != null) c.notes.push(`Fontes divergem ${Math.round(c.spreadPct * 100)}%: confira as condições (autoconexão/bilhetes separados).`);
  return c;
}
